### PR TITLE
fix: extension secrets should not be lost when workspace restart

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
### What does this PR do?
Allows to keep encrypted extension secrets in the browser local storage.

Screenshot from 2024-05-15 09-53-39

Launcher output
```
# Injecting server public key to che-code...
  > foud key file /etc/ssh/default-1611150837256.pub
  > out/vs/code/browser/workbench/workbench.js has been updated
```

### What issues does this PR fix?

https://github.com/eclipse-che/che/issues/22837

### How to test this PR?

1. Create a workspace with
https://github.com/vitaliy-guliy/vscode-test-extension/tree/persisting-secrets?che-editor=https://gist.githubusercontent.com/vitaliy-guliy/4fedeef3dbf8e902af94240780c442e6/raw/f08e91485b8529c6761b73f375fccde8d62ae94f/editor.yaml&new

2. Build test extension from sources or download the binary from https://github.com/vitaliy-guliy/vscode-test-extension/releases/tag/0.0.2

3. Go to extensions view, install the extension from .vsix file

4. Run `Secrets: Add secret` command and add the secret entering any value

5. Delete the workspace and create it again with the link above, or just restart the workspace (chose what do you prefer). If you recreated workspace, it needs to install test extension again.

6. Run `Secrets: Get Secret` command. You should get the value entered before.

You can create any workspace using `che-code` image provided by this pull request and experiment with any extension you want. For instance, it is possible to install sonarlint extension, open extension settings, configure it and then restart the workspace. All the settings should be saved.

### Does this PR contain changes that override default upstream Code-OSS behavior?
- [x] the PR contains changes in the [code](https://github.com/che-incubator/che-code/tree/main/code) folder (you can skip it if your changes are placed in a che extension )
- [x] the corresponding items were added to the [CHANGELOG.md](https://github.com/che-incubator/che-code/blob/main/.rebase/CHANGELOG.md) file
- [x] rules for automatic `git rebase` were added to the [.rebase](https://github.com/che-incubator/che-code/tree/main/.rebase) folder

### CODE REVIEW

**Review by azatsarynnyy — APPROVED** [review]
@vitaliy-guliy thanks for providing the test projects and steps for testing the PR.
I tested the PR following the suggested instructions. In a new workspace, I was able to get the secret, stored in another workspace.
👍

**Review by svor — APPROVED** [review]
Works good with the  Ansible extension

https://github.com/che-incubator/che-code/assets/1271546/148884d3-a161-4457-8825-13029bb4746f

File: launcher/src/local-storage-key-provider.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/**********************************************************************
+ * Copyright (c) 2023 Red Hat, Inc.
```
Comment (svor) [inline launcher/src/local-storage-key-provider.ts]:
```suggestion
 * Copyright (c) 2024 Red Hat, Inc.
```

File: launcher/src/local-storage-key-provider.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+const SERVER_KEY_MASK = '{{LOCAL-STORAGE}}/{{SECURE-KEY}}';
+
+const CERTS_DIR = '/etc/ssh';
+
+/**
+ * Finds a public key in `/etc/ssh` and initializes VS Code with 32 bytes (every fourth character) of the key.
+ * The key is used to encrypt/decrypt the extension secrets stored in browser local storage.
+ */
+export class LocalstorageKeyProvider {
```
Comment (svor) [inline launcher/src/local-storage-key-provider.ts]:
```suggestion
export class LocalStorageKeyProvider {
```

File: launcher/src/local-storage-key-provider.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  async configure(): Promise<void> {
+    console.log('# Injecting server public key to che-code...');
+
+    try {
+      const publicKeyFile = await this.findPublicKeyFile();
+      console.log(`  > foud key file ${publicKeyFile}`);
```
Comment (svor) [inline launcher/src/local-storage-key-provider.ts]:
```suggestion
      console.log(`  > found key file ${publicKeyFile}`);
```

File: launcher/src/local-storage-key-provider.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    return secret;
+  }
+
+  async update(file: string, text: string, newText: string): Promise<void> {
+    const content = await fs.readFile(file);
+    const newContent = content.replace(text, newText);
+
+    if (content === newContent) {
+      console.log(`  > ${file} is not updated`);
```
Comment (svor) [inline launcher/src/local-storage-key-provider.ts:75]: do we need these logs in the console?

File: launcher/src/main.ts [hunk 5]
```diff
@@ -12,6 +12,7 @@ import { CodeWorkspace } from './code-workspace';
 import { DevWorkspaceId } from './devworkspace-id';
 import { NodeExtraCertificate } from './node-extra-certificate';
 import { OpenVSIXRegistry } from './openvsix-registry';
+import { LocalstorageKeyProvider } from './local-storage-key-provider';
```
Comment (svor) [inline launcher/src/main.ts]:
```suggestion
import { LocalStorageKeyProvider } from './local-storage-key-provider';
```

File: launcher/src/main.ts [hunk 6]
```diff
@@ -28,6 +29,7 @@ export class Main {
     await new OpenVSIXRegistry().configure();
     await new WebviewResources().configure();
     await new NodeExtraCertificate().configure();
+    await new LocalstorageKeyProvider().configure();
```
Comment (svor) [inline launcher/src/main.ts]:
```suggestion
    await new LocalStorageKeyProvider().configure();
```

File: launcher/tests/local-storage-key-provider.spec.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
+/**********************************************************************
+ * Copyright (c) 2023 Red Hat, Inc.
```
Comment (svor) [inline launcher/tests/local-storage-key-provider.spec.ts]:
```suggestion
 * Copyright (c) 2024 Red Hat, Inc.
```

File: launcher/tests/local-storage-key-provider.spec.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+import * as fs from '../src/fs-extra';
+import { LocalstorageKeyProvider } from '../src/local-storage-key-provider';
```
Comment (svor) [inline launcher/tests/local-storage-key-provider.spec.ts]:
```suggestion
import { LocalStorageKeyProvider } from '../src/local-storage-key-provider';
```

File: launcher/tests/local-storage-key-provider.spec.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    readFileMock.mockImplementation(async (file: string) => {
+      switch (file) {
+        case '/etc/ssh/first-key.pub':
+          return 'ssh-rsa 1111222233334444555566667777888899990000AAAABBBBCCCCDDDDEEEEFFFFGGGGHHHHIIIIJJJJKKKKLLLLMMMMNNNNOOOOPPPPQQQQRRRRSSSSTTTTUUUUVVVVWWWWXXXXYYYYZZZZ';
+        case 'out/vs/code/browser/workbench/workbench.js':
+          return ORIGIN_WORKBENCH_FILE;
+      }
+    });
+
+    const localStorageKeyProvider = new LocalstorageKeyProvider();
```
Comment (svor) [inline launcher/tests/local-storage-key-provider.spec.ts]:
```suggestion
    const localStorageKeyProvider = new LocalStorageKeyProvider();
```

File: launcher/tests/main.spec.ts [hunk 10]
```diff
@@ -38,6 +38,13 @@ jest.mock('../src/node-extra-certificate', () => ({
   },
 }));
 
+const configureLocalstorageKeyProvider = jest.fn();
```
Comment (svor) [inline launcher/tests/main.spec.ts]:
```suggestion
const configureLocalStorageKeyProvider = jest.fn();
```

File: launcher/tests/main.spec.ts [hunk 11]
```diff
@@ -38,6 +38,13 @@ jest.mock('../src/node-extra-certificate', () => ({
...
+jest.mock('../src/local-storage-key-provider', () => ({
+  LocalstorageKeyProvider: function () {
```
Comment (svor) [inline launcher/tests/main.spec.ts]:
```suggestion
  LocalStorageKeyProvider: function () {
```

File: launcher/tests/main.spec.ts [hunk 12]
```diff
@@ -38,6 +38,13 @@ jest.mock('../src/node-extra-certificate', () => ({
...
+    return { configure: configureLocalstorageKeyProvider };
```
Comment (svor) [inline launcher/tests/main.spec.ts]:
```suggestion
    return { configure: configureLocalStorageKeyProvider };
```

File: launcher/tests/main.spec.ts [hunk 13]
```diff
@@ -67,6 +74,7 @@ describe('Test main flow:', () => {
     expect(configureOpenVSIXRegistryMock).toBeCalled();
     expect(configureWebviewResourcesMock).toBeCalled();
     expect(configureNodeExtraCertificate).toBeCalled();
+    expect(configureLocalstorageKeyProvider).toBeCalled();
```
Comment (svor) [inline launcher/tests/main.spec.ts]:
```suggestion
    expect(configureLocalStorageKeyProvider).toBeCalled();
```

File: launcher/src/local-storage-key-provider.ts (hunk 3 above)
Comment (svor) [reply launcher/src/local-storage-key-provider.ts]: do we really need these logs?

**Review by RomanNikitenko — APPROVED** [review]

File: launcher/src/local-storage-key-provider.ts (hunk 4 above)
Comment (vitaliy-guliy) [reply launcher/src/local-storage-key-provider.ts:75]: All the output is forwarded to the file `/checode/entrypoint-logs.txt`. You can open it in the editor and check the process of configuring/adaptation of VS Code to be running in Che.

**Review by ibuziuk — APPROVED** [review]

**Comment by musienko-maxim** [comment]
Verification completed from my side

**Pull request merged**